File: examples/agent-middleware/python/interai_middleware.py
import json
import os
import time
import urllib.error
import urllib.request
from typing import Any, Callable, Dict, Optional

AgentAction = Dict[str, Any]
VerifyResponse = Dict[str, Any]
ExecutionDecision = Dict[str, Any]


def _base_url() -> str:
    return os.environ.get("INTERAI_BASE_URL", "https://ai-risk-oracle.fly.dev").rstrip("/")


def _api_key() -> str:
    return os.environ.get("INTERAI_API_KEY", "")


def _receipt_id(response: VerifyResponse) -> Optional[str]:
    receipt = response.get("trust_receipt")
    if isinstance(receipt, dict) and isinstance(receipt.get("receipt_id"), str):
        return receipt["receipt_id"]
    value = response.get("trust_receipt_id")
    return value if isinstance(value, str) else None


def _validate_verify_response(value: Any) -> VerifyResponse:
    if not isinstance(value, dict):
        raise RuntimeError("Malformed InterAI response: expected JSON object")
    if value.get("recommended_action") not in ("allow", "review_required", "block"):
        raise RuntimeError("Malformed InterAI response: missing recommended_action")
    return value


def verify_before_execution(action: AgentAction) -> VerifyResponse:
    api_key = _api_key()
    if not api_key:
        raise RuntimeError("Missing INTERAI_API_KEY")

    payload = {
        "use_case": "agent-before-tool-execution",
        "action": action,
        "context": {
            "agent_id": "example_agent_python",
            "environment": "sandbox",
            "user_confirmation": False,
        },
        "policy": {
            "max_risk_level": "medium",
            "require_trust_receipt": True,
            "amount_usd_limit": 500,
            "blocked_action_types": ["irreversible_transfer"],
            "require_human_review_above": 0.75,
        },
    }

    request = urllib.request.Request(
        f"{_base_url()}/verify",
        data=json.dumps(payload).encode("utf-8"),
        headers={
            "content-type": "application/json",
            "authorization": f"Bearer {api_key}",
            "x-idempotency-key": f"agent-middleware-python-{int(time.time() * 1000)}",
        },
        method="POST",
    )

    try:
        with urllib.request.urlopen(request, timeout=15) as response:
            text = response.read().decode("utf-8")
            return _validate_verify_response(json.loads(text))
    except urllib.error.HTTPError as error:
        body = error.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"InterAI /verify failed: HTTP {error.code} {body[:300]}") from error
    except urllib.error.URLError as error:
        raise RuntimeError(f"Network failure calling InterAI: {error}") from error
    except json.JSONDecodeError as error:
        raise RuntimeError("Malformed InterAI response: non-JSON body") from error


def execute_with_interai_gate(
    action: AgentAction,
    executor: Callable[[AgentAction], Any],
) -> ExecutionDecision:
    verification = verify_before_execution(action)
    recommended_action = verification["recommended_action"]
    base = {
        "recommended_action": recommended_action,
        "risk_level": verification.get("risk_level"),
        "policy_result": verification.get("policy_result"),
        "trust_receipt_id": _receipt_id(verification),
        "verification": verification,
    }

    print(
        "InterAI decision",
        {
            "recommended_action": base["recommended_action"],
            "risk_level": base["risk_level"],
            "policy_result": base["policy_result"],
            "trust_receipt_id": base["trust_receipt_id"],
        },
    )

    if recommended_action == "allow":
        return {
            "status": "executed",
            **base,
            "result": executor(action),
        }

    if recommended_action == "review_required":
        return {
            "status": "review_required",
            **base,
            "review_reason": "InterAI requires supervisor or policy review before execution",
        }

    return {
        "status": "blocked",
        **base,
        "block_reason": "InterAI blocked this action before execution",
    }

File: examples/agent-middleware/typescript/interaiMiddleware.ts
export type AgentAction = {
  type: string
  name: string
  description: string
  external_side_effect: boolean
  irreversible: boolean
  amount_usd?: number
  currency?: string
  metadata?: Record<string, unknown>
}

export type VerifyRequest = {
  use_case: string
  action: AgentAction
  context: {
    agent_id: string
    environment: "sandbox" | "staging" | "production"
    user_confirmation: boolean
    counterparty_id?: string
  }
  policy: {
    max_risk_level: "low" | "medium" | "high"
    require_trust_receipt: boolean
    amount_usd_limit?: number
    blocked_action_types?: string[]
    require_human_review_above?: number
  }
}

export type VerifyResponse = {
  recommended_action?: "allow" | "review_required" | "block" | string
  risk_level?: "low" | "medium" | "high" | string
  policy_result?: "allow" | "review_required" | "block" | string
  trust_receipt_id?: string
  trust_receipt?: {
    receipt_id?: string
  }
  [key: string]: unknown
}

export type InterAIClientOptions = {
  apiKey?: string
  baseUrl?: string
  agentId?: string
  environment?: "sandbox" | "staging" | "production"
}

export type ExecutionDecision =
  | {
      status: "executed"
      recommended_action: "allow"
      risk_level: string | null
      policy_result: string | null
      trust_receipt_id: string | null
      result: unknown
      verification: VerifyResponse
    }
  | {
      status: "review_required"
      recommended_action: "review_required"
      risk_level: string | null
      policy_result: string | null
      trust_receipt_id: string | null
      review_reason: string
      verification: VerifyResponse
    }
  | {
      status: "blocked"
      recommended_action: "block"
      risk_level: string | null
      policy_result: string | null
      trust_receipt_id: string | null
      block_reason: string
      verification: VerifyResponse
    }

export type ActionExecutor<T = unknown> = (action: AgentAction) => Promise<T>

function clientOptions(options: InterAIClientOptions = {}) {
  return {
    apiKey: options.apiKey ?? process.env.INTERAI_API_KEY,
    baseUrl:
      options.baseUrl ??
      process.env.INTERAI_BASE_URL ??
      "https://ai-risk-oracle.fly.dev",
    agentId: options.agentId ?? "example_agent_middleware",
    environment: options.environment ?? "sandbox"
  }
}

function receiptIdFrom(response: VerifyResponse) {
  return response.trust_receipt_id ?? response.trust_receipt?.receipt_id ?? null
}

function assertValidVerifyResponse(value: unknown): asserts value is VerifyResponse {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error("Malformed InterAI response: expected JSON object")
  }

  const action = (value as VerifyResponse).recommended_action
  if (action !== "allow" && action !== "review_required" && action !== "block") {
    throw new Error("Malformed InterAI response: missing recommended_action")
  }
}

export async function verifyBeforeExecution(
  action: AgentAction,
  options: InterAIClientOptions = {}
): Promise<VerifyResponse> {
  const config = clientOptions(options)
  if (!config.apiKey) {
    throw new Error("Missing INTERAI_API_KEY")
  }

  const request: VerifyRequest = {
    use_case: "agent-before-tool-execution",
    action,
    context: {
      agent_id: config.agentId,
      environment: config.environment,
      user_confirmation: false
    },
    policy: {
      max_risk_level: "medium",
      require_trust_receipt: true,
      amount_usd_limit: 500,
      blocked_action_types: ["irreversible_transfer"],
      require_human_review_above: 0.75
    }
  }

  let response: Response
  try {
    response = await fetch(`${config.baseUrl.replace(/\/+$/, "")}/verify`, {
      method: "POST",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${config.apiKey}`,
        "x-idempotency-key": `agent-middleware-${Date.now()}`
      },
      body: JSON.stringify(request)
    })
  } catch (error) {
    throw new Error(
      `Network failure calling InterAI: ${
        error instanceof Error ? error.message : String(error)
      }`
    )
  }

  const text = await response.text()
  let body: unknown
  try {
    body = text ? JSON.parse(text) : {}
  } catch {
    throw new Error(`Malformed InterAI response: non-JSON HTTP ${response.status}`)
  }

  if (!response.ok) {
    throw new Error(
      `InterAI /verify failed: HTTP ${response.status} ${JSON.stringify(body).slice(
        0,
        300
      )}`
    )
  }

  assertValidVerifyResponse(body)
  return body
}

export async function executeWithInterAIGate<T>(
  action: AgentAction,
  executor: ActionExecutor<T>,
  options: InterAIClientOptions = {}
): Promise<ExecutionDecision> {
  const verification = await verifyBeforeExecution(action, options)
  const recommendedAction = verification.recommended_action
  const base = {
    risk_level: verification.risk_level ?? null,
    policy_result: verification.policy_result ?? null,
    trust_receipt_id: receiptIdFrom(verification),
    verification
  }

  console.log("InterAI decision", {
    recommended_action: recommendedAction,
    risk_level: base.risk_level,
    policy_result: base.policy_result,
    trust_receipt_id: base.trust_receipt_id
  })

  if (recommendedAction === "allow") {
    const result = await executor(action)
    return {
      status: "executed",
      recommended_action: "allow",
      ...base,
      result
    }
  }

  if (recommendedAction === "review_required") {
    return {
      status: "review_required",
      recommended_action: "review_required",
      ...base,
      review_reason: "InterAI requires supervisor or policy review before execution"
    }
  }

  return {
    status: "blocked",
    recommended_action: "block",
    ...base,
    block_reason: "InterAI blocked this action before execution"
  }
}
